Mathlib/FieldTheory/IsPerfectClosure.lean

```diff
@@ -0,0 +1,540 @@
+  rcases Nat.le_one_iff_eq_zero_or_eq_one.1 (not_lt.1 hp) with hp | hp
+  · by_cases hn : n = 0
+    · rwa [hn, pow_zero, pow_one] at h
+    rw [hp, zero_pow hn, pow_zero] at h
+    haveI := subsingleton_of_zero_eq_one h.symm
+    exact Subsingleton.elim _ _
+  rwa [hp, one_pow, pow_one] at h
+
+section IsPerfectClosure
+
+variable {K L M : Type*}
+
+section CommSemiring
+
+variable [CommSemiring K] [CommSemiring L] [CommSemiring M]
+  (i : K →+* L) (j : K →+* M) (p : ℕ) [ExpChar K p] [ExpChar L p] [ExpChar M p]
+
+/-- If `i : K →+* L` is a ring homomorphism of characteristic `p` rings, then it is called
+`p`-radical if the following conditions are satisfied:
+
+- For any element `x` of `L` there is `n : ℕ` such that `x ^ (p ^ n)` is contained in `K`.
+- The kernel of `i` is contained in the `p`-nilradical of `K`.
+
+It is a generalization of purely inseparable extension for fields. -/
+@[mk_iff]
+class IsPRadical : Prop where
+  pow_mem' : ∀ x : L, ∃ (n : ℕ) (y : K), i y = x ^ p ^ n
+  ker_le' : RingHom.ker i ≤ pNilradical K p
+
+theorem IsPRadical.pow_mem [IsPRadical i p] (x : L) :
+    ∃ (n : ℕ) (y : K), i y = x ^ p ^ n := pow_mem' x
+
+theorem IsPRadical.ker_le [IsPRadical i p] :
+    RingHom.ker i ≤ pNilradical K p := ker_le'
+
+theorem IsPRadical.comap_pNilradical [IsPRadical i p] :
+    (pNilradical L p).comap i = pNilradical K p := by
+  refine le_antisymm (fun x h ↦ mem_pNilradical.2 ?_) (fun x h ↦ ?_)
+  · obtain ⟨n, h⟩ := mem_pNilradical.1 <| Ideal.mem_comap.1 h
+    obtain ⟨m, h⟩ := mem_pNilradical.1 <| ker_le i p ((map_pow i x _).symm ▸ h)
+    exact ⟨n + m, by rwa [pow_add, pow_mul]⟩
+  simp only [Ideal.mem_comap, mem_pNilradical] at h ⊢
+  obtain ⟨n, h⟩ := h
+  exact ⟨n, by simpa only [map_pow, map_zero] using congr(i $h)⟩
+
+variable (K) in
+instance IsPRadical.of_id : IsPRadical (RingHom.id K) p where
+  pow_mem' x := ⟨0, x, by rw [RingHom.id_apply, pow_zero, pow_one]⟩
+  ker_le' x h := by
+    rw [RingHom.mem_ker, RingHom.id_apply] at h
+    exact h ▸ Ideal.zero_mem _
+
+theorem IsPRadical.trans (j : L →+* M) [IsPRadical i p] [IsPRadical j p] :
+    IsPRadical (j.comp i) p where
+  pow_mem' x := by
+    obtain ⟨n, y, hy⟩ := pow_mem j p x
+    obtain ⟨m, z, hz⟩ := pow_mem i p y
+    exact ⟨n + m, z, by rw [RingHom.comp_apply, hz, map_pow, hy, pow_add, pow_mul]⟩
+  ker_le' x h := by
+    rw [RingHom.mem_ker, RingHom.comp_apply, ← RingHom.mem_ker] at h
+    simpa only [← Ideal.mem_comap, comap_pNilradical] using ker_le j p h
+
+/-- If `i : K →+* L` is a ring homomorphism of characteristic `p` rings, then it makes `L`
+a perfect closure of `K` if the following conditions are satisfied:
+
+- `L` is a perfect ring.
+- `i : K →+* L` is `p`-radical.
+
+In this case the kernel of `i` is equal to the `p`-nilradical of `K`
+(see `IsPerfectClosure.ker_eq`). -/
+@[mk_iff]
+class IsPerfectClosure : Prop where
+  [perfectRing' : PerfectRing L p]
+  [isPRadical' : IsPRadical i p]
```

Comment:
Currently `PerfectRing + IsPRadical => IsPerfectClosure` and `IsPerfectClosure => IsPRadical` can be inferred automatically, but not `IsPerfectClosure => PerfectRing`. This annoys me a little bit.

I have an idea that we should not talk about the compound `IsPerfectClosure`, but talk about `PerfectRing + IsPRadical` everywhere. Any opinions?

Reply:
That would be at odds with the design of IsAlgClosure, IsSepClosure etc. and the bundled version may have performance benefits, but I think it's good to experiment with alternative designs!

Reply:
Oh, the same problem also occurred for `IsAlgClosure`, and `IsSepClosure`. Perhaps I should not change it for `IsPerfectClosure`.